### ISSUE TITLE
Suggestion: highlight member types when clicking on namespace (and vice versa).

### ISSUE THREAD

**Issue by ericmutta** [dotnet/roslyn#52993]
I am a recent (happy!) convert to C# (previously used VB for 20+ years) and since C# requires you to repeat namespace imports in every file (I really miss project-level imports from VB) I often find myself asking **why did I import this namespace?**

**Currently I have to comment out the line with the `using` declaration then navigate through the lines that break as a result to see which types I was using from the commented out namespace**. The compiler/IDE already supports the **highlight references** feature under `Tools > Options > C# > Advanced > Highlighting`:

![image](https://user-images.githubusercontent.com/20465797/116448657-0808fc00-a862-11eb-88d3-1188171aafcb.png)

I would like to suggest that this feature be **extended to types and the imported namespaces they come from**. For example, given this import declaration:

```csharp
using System.Threading.Tasks;
```

...and later down the file, code like this:

```csharp
 public async Task<ActionResult> LoginAsync(){}
```

...**I would like the IDE to highlight the `using` declaration when I put the cursor on the `Task` type name (and vice versa)**. This makes it easy to answer the question **why did I import this namespace?** by simply clicking on the `using` declaration then scrolling down the file to see which types are highlighted.

As a further enhancement to achieve the same goal, perhaps **the tooltip that's shown when you hover over a `using` declaration can be expanded to list some of the types used from that namespace**. At the moment it simply repeats what you can already see on the `using` declaration which is not very useful:

![image](https://user-images.githubusercontent.com/20465797/116450052-82864b80-a863-11eb-9e41-c0aa0fbe2941.png)

**Comment by ericmutta** [comment]
Thanks @jinujoseph for the super fast triaging of this suggestion, much appreciated :pray:

**Comment by CyrusNajmabadi** [comment]
I like both of these ideas.  Costs might be significant though.  For example, there's no cheap way to answer the question "what types/extension-methods are used from this `using directive`?".

I think "highlight refs" would be cheap to update though.

**Comment by ericmutta** [comment]
Thanks @jasonmalinowski and @JoeRobich for following up on this and providing updates :pray: 

I **especially like the idea of integrating CodeLens because you get a "hands-off" experience where you don't have to touch the mouse** in order to hover over or click on the directive - the information about its references is always there on screen :+1:

Ideally, the **CodeLens integration should do more than just showing the number of references**. When looking at a using directive I believe most people want to know two things:

1. **Is this directive required at all?** We already get this answered today: if a directive isn't required the editor fades out the text. If CodeLens was made to just show reference counts, it would show `0 references` in this case which is duplication of effort since the faded text already says as much.
2. **If the directive is required, why is it required?** This essentially asks what types (or extension methods) are being used and to get a good "hands-off" experience, the CodeLens should show this information.

In my experience, when asking question number 2 above, I **don't need to know _all_ of the types used, just one or two** to jog my memory as it were. To that end, here is a possible mock-up of how CodeLens could show information:

```csharp
// 3 types (List, Dictionary, Stack)  <-- CodeLens showing number of types and a listing of the first 3 types found.
using System.Collections.Generic;

// 1 method (ToList) <-- CodeLens showing number of extension methods and a listing of the only method found.
using System.Linq;

// 2 types (Doo, Dah) | 3 methods (Moo, Mah, Meh)  <-- CodeLens showing both types and extension methods.
using Foo.Bar.Baz;
```

Thanks in advance to everyone on the team that is following up on this suggestion. I know you've got a million other things to do, so the fact that you discussed this _at all_ is much appreciated :pray:

**Comment by jasonmalinowski** [comment]
@ericmutta: yep, all good thoughts. I think our general motto was once we have the API from the compiler layer to actually drive this then we'll start lighting it up wherever we can.

And there was very universal agreement we should do something -- usually people (myself included) work around this by deleting a using and seeing what breaks. Besides being slow it's also not entirely correct: there are edge cases in C# where a deleting a using can change overload resolution _without_ a type or method being consumed, so you could have code that changes meaning even if it's not obvious why.

**Comment by CyrusNajmabadi** [comment]
Closing out due to lack of feedback.
